wgpu-core/src/resource.rs

```diff
@@ -191,6 +192,33 @@ pub struct Texture<A: hal::Api> {
     pub(crate) life_guard: LifeGuard,
 }
 
+pub struct HalTextureGuard<'a, A: HalApi> {
+    guard: RwLockReadGuard<'a, Storage<Texture<A>, TextureId>>,
+    id: TextureId,
+}
+
+impl<'a, A: HalApi> HalTextureGuard<'a, A> {
+    pub fn as_inner(&'a self) -> Option<&'a A::Texture> {
+        let texture = self.guard.get(self.id).unwrap();
+
+        texture.inner.as_raw()
+    }
+}
+
+impl<G: GlobalIdentityHandlerFactory> Global<G> {
+    /// # Safety
+    ///
+    /// - The raw texture handle must not be manually destroyed
+    pub unsafe fn texture_as_hal<'a, A: HalApi>(&'a self, id: TextureId) -> HalTextureGuard<'a, A> {
```

Comment:
can this be shaped with a closure instead? i.e. function gets a closure that's ran over `&Texture`

Reply:
Thank you for the suggestion! This would remove the need for both `HalTextureGuard` and `Registry::unsafe_read()`.